tests/regression/src/lib.rs

```diff
@@ -105,52 +122,135 @@ mod tests {
+
+        println!("Running command: {:?}", command);
+        execute_command(command);
+
+        let annotate_output = run_cg_annotate(&output_file);
+        save_annotate_output(&annotate_output, suffix, test_name);
+
+        let count = find_instruction_count(&annotate_output)
+            .expect("Failed to get instruction count from file");
+
+        println!("Instruction count for {}: {}", test_name, count);
+    }
+
+    fn create_output_file_path(test_name: &str, suffix: &str) -> String {
         create_dir_all(Path::new("target/cg_artifacts")).unwrap();
-        let output_file = format!("target/cg_artifacts/cachegrind_{}.out", test_name);
-        let output_command = format!("--cachegrind-out-file={}", &output_file);
+        format!(
+            "target/cg_artifacts/cachegrind_{}_{}.out",
+            test_name, suffix
+        )
+    }
+
+    fn build_valgrind_command(exe_path: &str, test_name: &str, output_file: &str) -> Command {
+        let output_command = format!("--cachegrind-out-file={}", output_file);
         let mut command = Command::new("valgrind");
         command
-            .args(["--tool=cachegrind", &output_command, &exe_path, test_name])
-            // Ensures that the recursive call is made to the actual harness code block rather than back to this function
-            .env_remove("ENABLE_VALGRIND"); 
+            .args(["--tool=cachegrind", &output_command, exe_path, test_name])
+            .env_remove("ENABLE_VALGRIND");
+        command
+    }
 
-        println!("Running command: {:?}", command);
+    fn execute_command(mut command: Command) {
         let status = command.status().expect("Failed to execute valgrind");
-
         if !status.success() {
             panic!("Valgrind failed");
         }
+    }
 
+    fn run_cg_annotate(output_file: &str) -> std::process::Output {
         let annotate_output = Command::new("cg_annotate")
-            .arg(&output_file)
+            .arg(output_file)
             .output()
             .expect("Failed to run cg_annotate");
-
         if !annotate_output.status.success() {
             panic!("cg_annotate failed");
         }
-        create_dir_all(Path::new("target/perf_outputs")).unwrap();
-        let annotate_file = format!("target/perf_outputs/{}.annotated.txt", test_name);
-        let mut file = File::create(&annotate_file).expect("Failed to create annotation file");
-        file.write_all(&annotate_output.stdout)
+        annotate_output
+    }
+
+    /// Saves the annotated output to prev, curr, or diff accordingly
+    fn save_annotate_output(output: &std::process::Output, suffix: &str, test_name: &str) {
+        let directory = format!("target/perf_outputs/{}", suffix);
+        create_dir_all(Path::new(&directory)).unwrap();
+        let annotate_file = format!(
+            "target/perf_outputs/{}/{}_{}.annotated.txt",
+            suffix, test_name, suffix
```

Comment:
nit: it's a bit clearer to inline interpolation

```suggestion
            "target/perf_outputs/{suffix}/{test_name}_{suffix}.annotated.txt",
```

There's quite a few places in the code that could use this